glean-core/src/upload/mod.rs

```diff
@@ -288,19 +331,28 @@ impl PingUploadManager {
     ///
     /// ## Arguments
     ///
+    /// * `glean` - The Glean object holding the database.
     /// * `log_ping` - Whether to log the ping before returning.
     ///
     /// # Return value
     ///
     /// `PingUploadTask` - see [`PingUploadTask`](enum.PingUploadTask.html) for more information.
-    pub fn get_upload_task(&self, log_ping: bool) -> PingUploadTask {
+    pub fn get_upload_task(&self, glean: &Glean, log_ping: bool) -> PingUploadTask {
         if !self.has_processed_pings_dir() {
             log::info!(
                 "Tried getting an upload task, but processing is ongoing. Will come back later."
             );
             return PingUploadTask::Wait;
         }
 
+        let mut pending_pings = self
+            .pending_pings
+            .write()
+            .expect("Can't write to pending pings cache.");
+        while let Some((document_id, path, body, headers)) = pending_pings.pop() {
```

Comment:
`pop` takes the last element, wouldn't this result in the exact reversed order we'd like them in? Probably not that big of an impact.

If we want to iterate in order and drain the vector at the same time and ALSO don't want to waste processing time/allocations on shifting stuff around we can use the [`drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain) iterator:

```rust
for (document_id, path, body, headers) in pending_pings.drain(..) {
    // ...
}

Reply:
Oh you are right! Oops.